### PR TITLE
People not removed in an exclude should be able to post on the new epoch

### PR DESCRIPTION
**Pull request by Powersource**
Fixes https://github.com/ssbc/ssb-tribes2/issues/72

### CODE REVIEW

File: test/exclude-members.test.js [hunk 1]
```diff
@@ -154,3 +155,122 @@ test('add and remove a person, post on the new feed', async (t) => {
...
+      text: 'second post',
+      recps: [groupId],
+    })
+    .catch(t.fail)
+
+  t.notEquals(
+    secondFeedId,
+    firstFeedId,
+    'feed for second publish is different to first publish'
+  )
```
Comment (Powersource) [inline test/exclude-members.test.js:250]: This fails.

Checked replication and carol seems to catch up to the latest message on the feed that the re-addition is published to by alice, but only alice seems to find the re-addition in start(). Does that mean that carol doesn't manage to decrypt the addition? Still happens with 10 seconds timeout before carol publishes.

Comment (Powersource) [reply test/exclude-members.test.js:250]: in box2/monitor.js we need to change it to something like

```js
      //pull.filter(isForeignLeafFeedBranch),
      pull.asyncMap((leafBranch, cb) => {
        if (!rootKeys) return cb()
        if (isForeignLeafFeedBranch(leafBranch)) {
          const theirRoot = leafBranch[0]
          const theirLeaf = leafBranch[3]
          ssb.metafeeds.getTree(rootKeys.id, (err, myRoot) => {
```
with an else where we do the reverse case

Comment (Powersource) [reply test/exclude-members.test.js:250]: andré described the problem like:

The problem was that: Carol tries to create a triangle with Alice as soon as Carol discover's Alice's epoch 0 feed, BUT at that point Carol doesn't yet have an epoch 0 feed, so creating triangle fails, and Carol never tries it again.

However, Carol SHOULD create a triangle with Alice's epoch 0 feed as soon as Carol creates her own epoch 0 feed

Comment (Powersource) [reply test/exclude-members.test.js:250]: We're putting this bug in the "fridge" for now by putting all add-member msgs on the additions feed, something we were thinking about anyway. TODO make issue etc

Comment (Powersource) [reply test/exclude-members.test.js:250]: Andre got a task to update the exclusion spec. Should make issue for this repo

Comment (Powersource) [reply test/exclude-members.test.js:260]: Rewrote this issue to be just about moving stuff to the additions feed. Will work on that now #79

File: test/exclude-members.test.js [hunk 2]
```diff
@@ -154,3 +155,117 @@ test('add and remove a person, post on the new feed', async (t) => {
...
+    .publish({
+      type: 'test',
+      text: 'first post',
+      recps: [groupId],
+    })
+    .catch((err) => t.error(err, 'carol failed to publish on first feed'))
+
+  await alice.tribes2
+    .excludeMembers(groupId, [bobRoot.id])
+    .catch((err) => t.error(err, 'remove member fail'))
```
[the remark this answers is not in the document]
Comment (Powersource) [reply test/exclude-members.test.js]: Unresolving this, we should test this here once #79 is fixed

File: test/exclude-members.test.js [hunk 3]
```diff
@@ -174,3 +175,133 @@ test('add and remove a person, post on the new feed', async (t) => {
...
+    firstFeedId,
+    'feed for second publish is different to first publish'
+  )
+
+  const { writeKey: writeKey2 } = await carol.tribes2.get(groupId)
+
+  const branches = await pull(
+    carol.metafeeds.branchStream({ root: carolRoot, old: true, live: false }),
+    pull.collectAsPromise()
+  )
```
Comment (Powersource) [inline test/exclude-members.test.js:287]: after basing on #85 this

```js
  const branches = await pull(
    carol.metafeeds.branchStream({ root: carolRoot, old: true, live: false }),
    pull.collectAsPromise()
  )

  console.log('carol branches', branches)
```

logs 

```js
carol branches [
  [
    FeedDetails {
      id: [FeedDetails],
      parent: null,
      purpose: 'root',
      feedFormat: 'bendybutt-v1',
      seed: null,
      keys: null,
      recps: null,
      metadata: {},
      tombstoned: false,
      tombstoneReason: null
    }
  ]
]
```

even though carol has successfully posted twice to the group, including on different feeds

File: test/exclude-members.test.js (hunk 1 above)
Comment (Powersource) [reply test/exclude-members.test.js:280]: seems like this might work after merging #85 into this

File: test/exclude-members.test.js (hunk 3 above)
Comment (Powersource) [reply test/exclude-members.test.js:287]: realized that carolRoot is an object and not the root id. strange that it returns anything at all

Comment (Powersource) [reply test/exclude-members.test.js:287]: made related pr https://github.com/ssbc/ssb-meta-feeds/pull/122

File: test/exclude-members.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Powersource) [reply test/exclude-members.test.js]: booya works

File: index.js [hunk 4]
```diff
@@ -231,10 +231,11 @@ module.exports = {
                     // prettier-ignore
                     if (err) return cb(clarify(err, "Couldn't post init msg on new epoch when excluding members"))
 
-                    addMembers(groupId, remainingMembers, {}, (err) => {
+                    addMembers(groupId, remainingMembers, {}, (err, reMsg) => {
                       // prettier-ignore
                       if (err) return cb(clarify(err, "Couldn't re-add remaining members when excluding members"))
-                      return cb()
+                      //TODO: we probably don't want to return this exactly
+                      return cb(null, { excludeMsg, reAddMsg: reMsg })
```
Comment (Powersource) [inline index.js:238]: todo remove probably

File: index.js [hunk 5]
```diff
@@ -407,6 +408,78 @@ module.exports = {
         if (err) return cb(clarify(err, 'Error finding or creating additions feed when starting ssb-tribes2'))
         return cb()
       })
+
+      // look for new epochs that we're added to
+      ssb.metafeeds.findOrCreate((err, myRoot) => {
+        // prettier-ignore
+        if (err) return cb(clarify(err, 'todo'))
```
Comment (Powersource) [inline index.js:415]: clarify errs

File: index.js [hunk 6]
```diff
@@ -407,6 +408,78 @@ module.exports = {
...
+
+        pull(
+          ssb.db.query(
+            // TODO: does this output new stuff if we accept an invite to an old epoch and then find additions to newer epochs?
+            where(and(isDecrypted('box2'), type('group/add-member'))),
+            live({ old: true }),
+            toPullStream()
```
Comment (Powersource) [inline index.js:422]: make issue about creating test case for this

File: index.js [hunk 7]
```diff
@@ -407,6 +408,78 @@ module.exports = {
...
+          ),
+          // groups/epochs we're added to
+          pull.filter((msg) => {
+            console.log('found an addition', myRoot.id, msg.value.content.recps)
```
Comment (Powersource) [inline index.js:426]: remove all the console logs

File: index.js [hunk 8]
```diff
@@ -407,6 +408,78 @@ module.exports = {
...
+              ssb.box2.addGroupInfo(groupId, { key: newKey }, (err) => {
+                // prettier-ignore
+                if (err) return cb(clarify(err, 'todo'))
+
+                const newKeyPick = {
+                  key: newKey,
+                  scheme: keySchemes.private_group,
+                }
+                // TODO: naively guessing that this is the latest key for now
+                ssb.box2.pickGroupWriteKey(groupId, newKeyPick, (err) => {
```
Comment (Powersource) [inline index.js:463]: Make issue about making this logic more advanced

File: test/exclude-members.test.js [hunk 9]
```diff
@@ -174,3 +175,142 @@ test('add and remove a person, post on the new feed', async (t) => {
...
+    .then((res) => {
+      t.pass('alice excluded bob')
+      return res
+    })
+    .catch((err) => t.error(err, 'remove member fail'))
+
+  await replicate(alice, carol).catch(t.error)
+
+  // TODO: maybe remove?
+  await p(setTimeout)(10000)
```
Comment (Powersource) [inline test/exclude-members.test.js:257]: remove or shorten

File: index.js (hunk 4 above)
Comment (Powersource) [reply index.js]: removed

File: index.js (hunk 6 above)
Comment (Powersource) [reply index.js:421]: https://github.com/ssbc/ssb-tribes2/issues/86

File: index.js (hunk 8 above)
Comment (Powersource) [reply index.js:458]: https://github.com/ssbc/ssb-tribes2/issues/87

File: index.js [hunk 10]
```diff
@@ -407,6 +407,72 @@ module.exports = {
...
+                if (err) return cb(clarify(err, "Error getting groups we're already in when looking for new epochs"))
+
+                if (groupIds.includes(msg.value?.content?.recps?.[0])) {
+                  return cb(null, msg)
+                } else {
+                  return cb()
+                }
+              })
+            )
+          }),
```
Comment (mixmix) [inline index.js:442]: `paraMap(fn) === pull.asyncMap(fn)`
You need to add a "width" for it to be running in parallel e.g. run at most 4 at once:

```suggestion
          }, 4),
```

File: index.js [hunk 11]
```diff
@@ -407,6 +407,72 @@ module.exports = {
...
+                // prettier-ignore
+                if (err) return cb(clarify(err, "Error getting groups we're already in when looking for new epochs"))
+
+                if (groupIds.includes(msg.value?.content?.recps?.[0])) {
+                  return cb(null, msg)
+                } else {
+                  return cb()
+                }
+              })
+            )
```
Comment (mixmix) [inline index.js:441]: we can move faster by comparing as we go and stopping as soon as we find a match

```suggestion
            const groupId = msg.value?.content?.recps?.[0]
            if (!groupId) return cb(null, null)

            pull(
              ssb.box2.listGroupIds(),
              pull.filter(id => id === groupId),
              pull.take(1),
              pull.collect((err, groupIds) => {
                // prettier-ignore
                if (err) return cb(clarify(err, "Error getting groups we're already in when looking for new epochs"))

                if (groupIds.length) return cb(null, msg)
                else cb(null, null)
              })
            )
```

File: index.js [hunk 12]
```diff
@@ -407,6 +407,72 @@ module.exports = {
...
+          pull.drain(
+            (msg) => {
+              const groupId = msg.value?.content?.recps?.[0]
+
+              const newKey = Buffer.from(msg.value?.content?.groupKey, 'base64')
+              ssb.box2.addGroupInfo(groupId, { key: newKey }, (err) => {
+                // prettier-ignore
+                if (err) return cb(clarify(err, 'Error adding new epoch key that we found'))
+
+                const newKeyPick = {
```
Comment (mixmix) [inline index.js:453]: I suspect we may have to record the epoch root for each key. I don't think it's sufficient to just swap in the latest key as the  correct writeKey.
Ah you have a TODO about it being naive, nice

BTW I really like your PR style of "here is a bit" and clearly leaving TODOs which are for the next round. When I worked on tribes 1 I loaded everything up in my head and was like "it's all got to work" :facepalm:

File: index.js [hunk 13]
```diff
@@ -407,6 +407,72 @@ module.exports = {
...
+        pull(
+          ssb.db.query(
+            // TODO: does this output new stuff if we accept an invite to an old epoch and then find additions to newer epochs?
+            where(and(isDecrypted('box2'), type('group/add-member'))),
+            live({ old: true }),
+            toPullStream()
+          ),
+          // groups/epochs we're added to
+          pull.filter((msg) => {
+            return msg.value?.content?.recps?.includes(myRoot.id)
```
Comment (mixmix) [inline index.js:425]:
```suggestion
            return isGroupAdd(msg) && msg.value.content.recps.includes(myRoot.id)
```

**Review by mixmix — CHANGES_REQUESTED** [review]
one :fire: (must change)
minor performance suggestions

Good clear test <3

File: index.js [hunk 14]
```diff
@@ -407,6 +407,72 @@ module.exports = {
...
+                } else {
+                  return cb()
+                }
+              })
+            )
+          }),
+          pull.filter(Boolean),
+          pull.drain(
+            (msg) => {
+              const groupId = msg.value?.content?.recps?.[0]
```
[the remark this answers is not in the document]
Comment (Powersource) [reply index.js]: done

File: index.js (hunk 10 above)
Comment (Powersource) [reply index.js]: done

File: index.js (hunk 11 above)
Comment (Powersource) [reply index.js]: listGroupIds is basically sync, it's just a stream to allow for its live opt

File: index.js (hunk 13 above)
Comment (Powersource) [reply index.js]: checked against the spec in a filter before this so removed the ?s

**Review by mixmix — APPROVED** [review]

**Pull request merged**